# Welcome to a lab_🐼s!

Last week, you did amazing work in setting up your computer to do Data Science!  Let's now do Data Science on your computer for the first time!

A few tips to remember:

- Refer to your **lecture notebook** and the **pandas cheat sheet** to help you out with the code!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to communicate about your work.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to explain your code, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- Every lab you will need to complete some parts of the lab with your group.  There will be both "group discussions" and "individual reflections" as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [12]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Daniel Kiv"
ca1_name = "Nick Wong"
ca2_name = "Emily Crawford"


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "Jude"
partner1_netid = "judey2"
partner1_major = "Engineering Technology & Management"

partner2_name = "Christ"
partner2_netid = "christg2"
partner2_major = "Undeclared"

partner3_name = "Pedro Padilla"
partner3_netid = "ppadi3"
partner3_major = "Accountancy"

<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

Last week, we explored the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

- You will come up with questions to ask the class for the "Hello Dataset"!
- *(This was originally on lab_intro, so you can use your answers from there if you got to this part!  If you did not, work with your group right now!)*

## Hello Dataset Overview

Your lab homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **not be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*

**What is your binary question?**

*Do you have any pets?*

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**

*Pets tend to be something which people have split opinions over. Some people like keeping pets and some don't. Some surveys show that around 53% of Americans have pets. Keeping that in mind it is possible that the class has a 50/50 split*

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

*How many siblings do you have?*

**What do you think the most frequent answer will be?**

*1 sibling*

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentences on why you believe that it will be the most frequent answer.**

*From my personal experience, most people who I have met have 1 sibling as two is considered the ideal number of children*

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

*What is your favourite colour?*

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

*Two people voted for blue and two people voted for red*

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use one of the first three questions but you want to know the classes' answer?

**What is your final question for the whole class?**

*Do you use Windows or Mac?*

## Submit Your Questions to Your TA

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/p1HvVZFT1jeSWRKL9


<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In literally every lab, you will need to import the library just like we did in lecture -- import pandas as `pd` in the following cell:

In [3]:
# Write the code to import pandas as pd:
import pandas as pd

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

In [4]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv")

In [5]:
# Print out the contents of `df`:
print(df)

       Year  Term YearTerm Subject  Number                          Name  \
0      2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
1      2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
2      2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
3      2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
4      2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
...     ...   ...      ...     ...     ...                           ...   
12638  2022  Fall  2022-fa    YDSH     101           Beginning Yiddish I   
12639  2022  Fall  2022-fa    YDSH     104       Intermediate Yiddish II   
12640  2022  Fall  2022-fa    ZULU     201             Elementary Zulu I   
12641  2022  Fall  2022-fa    ZULU     202            Elementary Zulu II   
12642  2022  Fall  2022-fa    ZULU     403           Intermediate Zulu I   

                                             Description Credit Hours  \
0      Interdi

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [4]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 12643 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [5]:
df_STAT = df[df.Subject == "STAT"]
df_STAT

Year  Term YearTerm Subject  Number                           Name  \
11890  2022  Fall  2022-fa    STAT     100                     Statistics   
11891  2022  Fall  2022-fa    STAT     100                     Statistics   
11892  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11893  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11894  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
...     ...   ...      ...     ...     ...                            ...   
12025  2022  Fall  2022-fa    STAT     578           Topics in Statistics   
12026  2022  Fall  2022-fa    STAT     590  Individual Study and Research   
12027  2022  Fall  2022-fa    STAT     593                STAT Internship   
12028  2022  Fall  2022-fa    STAT     593                STAT Internship   
12029  2022  Fall  2022-fa    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11890  First course in probability and statistics at ...        3 hours.   
11891  First course in probability and statistics at ...        3 hours.   
11892  Data Science Discovery is the intersection of ...        4 hours.   
11893  Data Science Discovery is the intersection of ...        4 hours.   
11894  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
12025  May be repeated if topics vary. Prerequisite: ...        4 hours.   
12026  Directed reading and research. Approved for le...   0 TO 8 hours.   
12027  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12028  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12029  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11890  Credit is not given for both STAT 100 and any ...   
11891  Credit is not given for both STAT 100 and any ...   
11892                         Same as CS 107 and IS 107.   
11893                         Same as CS 107 and IS 107.   
11894                         Same as CS 107 and IS 107.   
...                                                  ...   
12025  May be repeated if topics vary. Prerequisite: ...   
12026  Approved for letter and S/U grading. May be re...   
12027  Approved for letter and S/U grading. Prerequis...   
12028  Approved for letter and S/U grading. Prerequis...   
12029  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11890  Quantitative Reasoning I course.  ...              A   
11891  Quantitative Reasoning I course.  ...              A   
11892  Quantitative Reasoning I course.  ...              A   
11893  Quantitative Reasoning I course.  ...              A   
11894  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
12025                               NaN  ...              A   
12026                               NaN  ...              A   
12027                               NaN  ...              A   
12028                               NaN  ...              A   
12029                               NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11890             Closed     Lecture-Discussion       LCD   12:00 PM   
11891             Closed                 Online       ONL   ARRANGED   
11892      CrossListOpen                Lecture       LEC   11:00 AM   
11893             Closed  Laboratory-Discussion       LBD   12:30 PM   
11894             Closed  Laboratory-Discussion       LBD   12:30 PM   
...                  ...                    ...       ...        ...   
12025  Open (Restricted)     Lecture-Discussion        S0   12:00 PM   
12026  Open (Restricted)      Independent Study       IND   ARRANGED   
12027  Open (Restricted)             Intern

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [6]:
df_CS = df[df.Subject == "CS"]
df_CS

Year  Term YearTerm Subject  Number                              Name  \
3349  2022  Fall  2022-fa      CS     100      Computer Science Orientation   
3350  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3351  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3352  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3353  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...   ...      ...     ...     ...                               ...   
3731  2022  Fall  2022-fa      CS     598                    Special Topics   
3732  2022  Fall  2022-fa      CS     598                    Special Topics   
3733  2022  Fall  2022-fa      CS     598                    Special Topics   
3734  2022  Fall  2022-fa      CS     598                    Special Topics   
3735  2022  Fall  2022-fa      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3349  Introduction to Computer Science as a field an...        1 hours.   
3350  Fundamental principles, concepts, and methods ...        3 hours.   
3351  Fundamental principles, concepts, and methods ...        3 hours.   
3352  Fundamental principles, concepts, and methods ...        3 hours.   
3353  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3731  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3732  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3733  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3734  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3735    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3349                                                NaN   
3350  Prerequisite: One of MATH 220 or MATH 221 or M...   
3351  Prerequisite: One of MATH 220 or MATH 221 or M...   
3352  Prerequisite: One of MATH 220 or MATH 221 or M...   
3353  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3731  May be repeated in the same or separate terms ...   
3732  May be repeated in the same or separate terms ...   
3733  May be repeated in the same or separate terms ...   
3734  May be repeated in the same or separate terms ...   
3735    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3349                                NaN  ...              A   
3350  Quantitative Reasoning II course.  ...              A   
3351  Quantitative Reasoning II course.  ...              A   
3352  Quantitative Reasoning II course.  ...              A   
3353  Quantitative Reasoning II course.  ...              P   
...                                 ...  ...            ...   
3731                                NaN  ...              A   
3732                                NaN  ...              A   
3733                                NaN  ...              A   
3734                                NaN  ...              A   
3735                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3349  Open (Restricted)                Lecture       LEC   03:00 PM  03:50 PM   
3350  Open (Restricted)     Lecture-Discussion       LCD   10:00 AM  10:50 AM   
3351               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3352               Open  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3353             Closed  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   
...                 ...                    ...       ...        ...       ...   
3731             Closed     Lecture-Discussion       LCD   11:00 AM  12:15 PM   
3732  Open (Restricted)     Lecture-Discus

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [7]:
df_IS = df[df.Subject == "IS"]
df_IS

Year  Term YearTerm Subject  Number  \
6663  2022  Fall  2022-fa      IS     100   
6664  2022  Fall  2022-fa      IS     100   
6665  2022  Fall  2022-fa      IS     100   
6666  2022  Fall  2022-fa      IS     100   
6667  2022  Fall  2022-fa      IS     100   
...    ...   ...      ...     ...     ...   
6907  2022  Fall  2022-fa      IS     596   
6908  2022  Fall  2022-fa      IS     597   
6909  2022  Fall  2022-fa      IS     597   
6910  2022  Fall  2022-fa      IS     597   
6911  2022  Fall  2022-fa      IS     599   

                                                   Name  \
6663   Exploring the iSchool with a Human-Centered Lens   
6664   Exploring the iSchool with a Human-Centered Lens   
6665   Exploring the iSchool with a Human-Centered Lens   
6666   Exploring the iSchool with a Human-Centered Lens   
6667   Exploring the iSchool with a Human-Centered Lens   
...                                                 ...   
6907  Advanced Topics in Human-Centered Design &amp;...   
6908  Advanced Topics in Data Analytics &amp; Data S...   
6909  Advanced Topics in Data Analytics &amp; Data S...   
6910  Advanced Topics in Data Analytics &amp; Data S...   
6911                                    Thesis Research   

                                            Description    Credit Hours  \
6663  This course introduces students to the School ...        1 hours.   
6664  This course introduces students to the School ...        1 hours.   
6665  This course introduces students to the School ...        1 hours.   
6666  This course introduces students to the School ...        1 hours.   
6667  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
6907  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6908  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6909  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6910  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6911  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
6663  Prerequisite: Restricted to Majors Only; First...               NaN   
6664  Prerequisite: Restricted to Majors Only; First...               NaN   
6665  Prerequisite: Restricted to Majors Only; First...               NaN   
6666  Prerequisite: Restricted to Majors Only; First...               NaN   
6667  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
6907  2 or 4 graduate hours. No professional credit....               NaN   
6908  2 or 4 graduate hours. No professional credit....               NaN   
6909  2 or 4 graduate hours. No professional credit....               NaN   
6910  2 or 4 graduate hours. No professional credit....               NaN   
6911  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status           Enrollment Status                Type  \
6663  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6664  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6665  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6666  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6667  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
...   ...            ...                         ...                 ...   
6907  ...              A           Open (Restricted)              Online   
6908  ...              A           Open (Restricted)  Lecture-Discussion   
6909  ...              A           Open (Restricted)  Lecture-Discussion   
6910  ...              A           Open (Restricted)              Online   
6911  ...              A           Open (Restricted)   Independent Study   

     Type Code St

### 🔬 Test Case Checkpoint 🔬


In [8]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?


*STAT offers the least number of courses (140) and CS offers the most courses (387)*

### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


*All three, STAT, CS, and IS, have the highest course number 599 which is Thesis Research. This is probably because a thesis is required at the end of your masters for all majors. They are also all 500 level courses as that is usually the default for masters classes.*

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!



### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [9]:
mydf = df[df.Subject == "CS"]
mydf

Year  Term YearTerm Subject  Number                              Name  \
3349  2022  Fall  2022-fa      CS     100      Computer Science Orientation   
3350  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3351  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3352  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3353  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...   ...      ...     ...     ...                               ...   
3731  2022  Fall  2022-fa      CS     598                    Special Topics   
3732  2022  Fall  2022-fa      CS     598                    Special Topics   
3733  2022  Fall  2022-fa      CS     598                    Special Topics   
3734  2022  Fall  2022-fa      CS     598                    Special Topics   
3735  2022  Fall  2022-fa      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3349  Introduction to Computer Science as a field an...        1 hours.   
3350  Fundamental principles, concepts, and methods ...        3 hours.   
3351  Fundamental principles, concepts, and methods ...        3 hours.   
3352  Fundamental principles, concepts, and methods ...        3 hours.   
3353  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3731  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3732  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3733  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3734  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3735    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3349                                                NaN   
3350  Prerequisite: One of MATH 220 or MATH 221 or M...   
3351  Prerequisite: One of MATH 220 or MATH 221 or M...   
3352  Prerequisite: One of MATH 220 or MATH 221 or M...   
3353  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3731  May be repeated in the same or separate terms ...   
3732  May be repeated in the same or separate terms ...   
3733  May be repeated in the same or separate terms ...   
3734  May be repeated in the same or separate terms ...   
3735    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3349                                NaN  ...              A   
3350  Quantitative Reasoning II course.  ...              A   
3351  Quantitative Reasoning II course.  ...              A   
3352  Quantitative Reasoning II course.  ...              A   
3353  Quantitative Reasoning II course.  ...              P   
...                                 ...  ...            ...   
3731                                NaN  ...              A   
3732                                NaN  ...              A   
3733                                NaN  ...              A   
3734                                NaN  ...              A   
3735                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3349  Open (Restricted)                Lecture       LEC   03:00 PM  03:50 PM   
3350  Open (Restricted)     Lecture-Discussion       LCD   10:00 AM  10:50 AM   
3351               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3352               Open  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3353             Closed  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   
...                 ...                    ...       ...        ...       ...   
3731             Closed     Lecture-Discussion       LCD   11:00 AM  12:15 PM   
3732  Open (Restricted)     Lecture-Discus

### Puzzle 4.2:

Select **five random courses** from your major, using `mydf` and pandas functions:

In [10]:
random = mydf.sample(n=5)
random

Year  Term YearTerm Subject  Number                    Name  \
3501  2022  Fall  2022-fa      CS     341      System Programming   
3397  2022  Fall  2022-fa      CS     107  Data Science Discovery   
3572  2022  Fall  2022-fa      CS     424       Real-Time Systems   
3472  2022  Fall  2022-fa      CS     225         Data Structures   
3715  2022  Fall  2022-fa      CS     598          Special Topics   

                                            Description   Credit Hours  \
3501  Basics of system programming, including POSIX ...       4 hours.   
3397         Same as IS 107 and STAT 107. See STAT 107.       4 hours.   
3572  Supervisory control aspects of Cyber Physical ...  3 OR 4 hours.   
3472  Data abstractions: elementary data structures ...       4 hours.   
3715  Subject offerings of new and developing areas ...  2 TO 4 hours.   

                                           Section Info  \
3501  Credit is not given for both CS 341 and either...   
3397         Same as IS 107 and STAT 107. See STAT 107.   
3572  3 undergraduate hours. 3 or 4 graduate hours. ...   
3472  Credit is not given for CS 277 if credit for C...   
3715  May be repeated in the same or separate terms ...   

                      Degree Attributes  ... Section Status  \
3501                                NaN  ...              A   
3397   Quantitative Reasoning I course.  ...              A   
3572                                NaN  ...              A   
3472  Quantitative Reasoning II course.  ...              A   
3715                                NaN  ...              A   

               Enrollment Status                   Type Type Code Start Time  \
3501                        Open  Laboratory-Discussion       LBD   05:00 PM   
3397                      Closed  Laboratory-Discussion       LBD   12:30 PM   
3572  CrossListOpen (Restricted)     Lecture-Discussion       LCD   12:30 PM   
3472                      Closed  Laboratory-Discussion       LBD   11:00 AM   
3715           Open (Restricted)     Lecture-Discussion        S1   03:30 PM   

      End Time Days of Week  Room                       Building  \
3501  06:20 PM            W   403     Siebel Center for Comp Sci   
3397  01:50 PM            R  3038  Campus Instructional Facility   
3572  01:45 PM           WF   216     Siebel Center for Comp Sci   
3472  12:50 PM            R  3101       Sidney Lu Mech Engr Bldg   
3715  04:45 PM           TR  1043       Sidney Lu Mech Engr Bldg   

              Instructors  
3501           Angrave, L  
3397               Das, T  
3572        Abdelzaher, T  
3472  Evans, C;Solomon, B  
3715           Ganesan, A  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered courses** from your major:

In [11]:
bottom = mydf.nsmallest(7, "Number")
bottom

Year  Term YearTerm Subject  Number                              Name  \
3349  2022  Fall  2022-fa      CS     100      Computer Science Orientation   
3350  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3351  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3352  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3353  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3354  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3355  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   

                                            Description Credit Hours  \
3349  Introduction to Computer Science as a field an...     1 hours.   
3350  Fundamental principles, concepts, and methods ...     3 hours.   
3351  Fundamental principles, concepts, and methods ...     3 hours.   
3352  Fundamental principles, concepts, and methods ...     3 hours.   
3353  Fundamental principles, concepts, and methods ...     3 hours.   
3354  Fundamental principles, concepts, and methods ...     3 hours.   
3355  Fundamental principles, concepts, and methods ...     3 hours.   

                                           Section Info  \
3349                                                NaN   
3350  Prerequisite: One of MATH 220 or MATH 221 or M...   
3351  Prerequisite: One of MATH 220 or MATH 221 or M...   
3352  Prerequisite: One of MATH 220 or MATH 221 or M...   
3353  Prerequisite: One of MATH 220 or MATH 221 or M...   
3354  Prerequisite: One of MATH 220 or MATH 221 or M...   
3355  Prerequisite: One of MATH 220 or MATH 221 or M...   

                      Degree Attributes  ... Section Status  \
3349                                NaN  ...              A   
3350  Quantitative Reasoning II course.  ...              A   
3351  Quantitative Reasoning II course.  ...              A   
3352  Quantitative Reasoning II course.  ...              A   
3353  Quantitative Reasoning II course.  ...              P   
3354  Quantitative Reasoning II course.  ...              A   
3355  Quantitative Reasoning II course.  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3349  Open (Restricted)                Lecture       LEC   03:00 PM  03:50 PM   
3350  Open (Restricted)     Lecture-Discussion       LCD   10:00 AM  10:50 AM   
3351               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3352               Open  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3353             Closed  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   
3354               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3355               Open  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   

     Days of Week   Room                     Building            Instructors  
3349            F  THEAT                 Lincoln Hall               Fleck, M  
3350           MW    AUD        Foellinger Auditorium  Beckman, M;Tahboub, R  
3351            M   L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3352            M   L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3353            M   L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3354            T   L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3355            T   L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered courses** from your major:

In [14]:
highest = mydf.nlargest(9, "Number")
highest

Year  Term YearTerm Subject  Number             Name  \
3735  2022  Fall  2022-fa      CS     599  Thesis Research   
3714  2022  Fall  2022-fa      CS     598   Special Topics   
3715  2022  Fall  2022-fa      CS     598   Special Topics   
3716  2022  Fall  2022-fa      CS     598   Special Topics   
3717  2022  Fall  2022-fa      CS     598   Special Topics   
3718  2022  Fall  2022-fa      CS     598   Special Topics   
3719  2022  Fall  2022-fa      CS     598   Special Topics   
3720  2022  Fall  2022-fa      CS     598   Special Topics   
3721  2022  Fall  2022-fa      CS     598   Special Topics   

                                            Description    Credit Hours  \
3735    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   
3714  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3715  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3716  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3717  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3718  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3719  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3720  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3721  Subject offerings of new and developing areas ...   2 TO 4 hours.   

                                           Section Info Degree Attributes  \
3735    Approved for S/U grading only. May be repeated.               NaN   
3714  May be repeated in the same or separate terms ...               NaN   
3715  May be repeated in the same or separate terms ...               NaN   
3716  May be repeated in the same or separate terms ...               NaN   
3717  May be repeated in the same or separate terms ...               NaN   
3718  May be repeated in the same or separate terms ...               NaN   
3719  May be repeated in the same or separate terms ...               NaN   
3720  May be repeated in the same or separate terms ...               NaN   
3721  May be repeated in the same or separate terms ...               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
3735  ...              A             Closed   Independent Study       IND   
3714  ...              A             Closed              Online        E1   
3715  ...              A  Open (Restricted)  Lecture-Discussion        S1   
3716  ...              A  Open (Restricted)  Lecture-Discussion       S11   
3717  ...              A  Open (Restricted)              Online        E4   
3718  ...              A  Open (Restricted)  Lecture-Discussion        S9   
3719  ...              A  Open (Restricted)             Lecture        L5   
3720  ...              A  Open (Restricted)  Lecture-Discussion        S0   
3721  ...              A  Open (Restricted)  Lecture-Discussion        S8   

     Start Time  End Time Days of Week  Room                    Building  \
3735   ARRANGED       NaN          NaN   NaN                         NaN   
3714   ARRANGED       NaN          NaN   NaN                         NaN   
3715   03:30 PM  04:45 PM           TR  1043    Sidney Lu Mech Engr Bldg   
3716   02:00 PM  03:15 PM           TR  1103  Siebel Center for Comp Sci   
3717   ARRANGED       NaN          NaN   NaN                         NaN   
3718   09:30 AM  10:45 AM           TR  1105  Siebel Center for Comp Sci   
3719   03:30 PM  04:45 PM           TR  1105  Siebel Center for Comp Sci   
3720   02:00 PM  03:15 PM           WF  1302  Siebel Center for Comp Sci   
3721   11:00 AM  12:15 PM           TR  1302  Siebel Center for Comp Sci   

           Instructors  
3735               NaN  
3714         Renear, A  
3715        Ganesan, A  
3716  Ray Chaudhury, B  
3717        Farivar, R  
3718         Mendis, C  
3719       Delgosha, P  
3720          Heath, D  
3721    Soltanaghai, E  

[9 rows x 27 columns]

### Puzzle 4.5:

Select **a random sample of 10% of the courses** from your major:

In [15]:
r_sample = mydf.sample(frac=0.1)
r_sample

Year  Term YearTerm Subject  Number  \
3545  2022  Fall  2022-fa      CS     409   
3419  2022  Fall  2022-fa      CS     124   
3440  2022  Fall  2022-fa      CS     128   
3406  2022  Fall  2022-fa      CS     124   
3439  2022  Fall  2022-fa      CS     128   
3692  2022  Fall  2022-fa      CS     562   
3658  2022  Fall  2022-fa      CS     491   
3356  2022  Fall  2022-fa      CS     101   
3511  2022  Fall  2022-fa      CS     361   
3495  2022  Fall  2022-fa      CS     341   
3443  2022  Fall  2022-fa      CS     128   
3635  2022  Fall  2022-fa      CS     465   
3639  2022  Fall  2022-fa      CS     468   
3606  2022  Fall  2022-fa      CS     446   
3548  2022  Fall  2022-fa      CS     410   
3641  2022  Fall  2022-fa      CS     470   
3375  2022  Fall  2022-fa      CS     105   
3681  2022  Fall  2022-fa      CS     534   
3676  2022  Fall  2022-fa      CS     500   
3687  2022  Fall  2022-fa      CS     546   
3665  2022  Fall  2022-fa      CS     498   
3446  2022  Fall  2022-fa      CS     128   
3604  2022  Fall  2022-fa      CS     446   
3447  2022  Fall  2022-fa      CS     128   
3425  2022  Fall  2022-fa      CS     124   
3534  2022  Fall  2022-fa      CS     374   
3525  2022  Fall  2022-fa      CS     374   
3384  2022  Fall  2022-fa      CS     107   
3709  2022  Fall  2022-fa      CS     591   
3701  2022  Fall  2022-fa      CS     591   
3591  2022  Fall  2022-fa      CS     437   
3390  2022  Fall  2022-fa      CS     107   
3486  2022  Fall  2022-fa      CS     233   
3618  2022  Fall  2022-fa      CS     461   
3481  2022  Fall  2022-fa      CS     233   
3465  2022  Fall  2022-fa      CS     199   
3460  2022  Fall  2022-fa      CS     199   
3455  2022  Fall  2022-fa      CS     199   
3514  2022  Fall  2022-fa      CS     361   

                                                   Name  \
3545                         The Art of Web Programming   
3419                 Introduction to Computer Science I   
3440                Introduction to Computer Science II   
3406                 Introduction to Computer Science I   
3439                Introduction to Computer Science II   
3692  Advanced Topics in Security, Privacy, and Mach...   
3658                                            Seminar   
3356                   Intro Computing: Engrg &amp; Sci   
3511  Probability &amp; Statistics for Computer Science   
3495                                 System Programming   
3443                Introduction to Computer Science II   
3635                              User Interface Design   
3639                     Tech and Advertising Campaigns   
3606                                   Machine Learning   
3548                           Text Information Systems   
3641                    Social and Information Networks   
3375                          Intro Computing: Non-Tech   
3681           Advanced Topics in Computer Architecture   
3676     Current Topics in Computing Education Research   
3687     Advanced Topics in Natural Language Processing   
3665                                     Special Topics   
3446                Introduction to Computer Science II   
3604                                   Machine Learning   
3447                Introduction to Computer Science II   
3425                 Introduction to Computer Science I   
3534  Introduction to Algorithms &amp; Models of Com...   
3525  Introduction to Algorithms &amp; Models of Com...   
3384                             Data Science Discovery   
3709                                   Advanced Seminar   
3701                                   Advanced Seminar   
3591                       Topics in Internet of Things   
3390                             Data Science Discovery   
3486                              Computer Architecture   
3618                                Computer Security I   
3481                              Computer Architecture   
3465     Undergraduate Open Seminar in Computer Science   
3460     Undergraduate

### ✔️ Checkpoint ✔️

There is no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is a new relatively new major here at UIUC.  The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [16]:
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the 2022 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [17]:
df2019_IS = df2019[df2019.Subject == "IS"]
df2019_IS

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [18]:
df_IS = df[df.Subject == "IS"]
df_IS

Year  Term YearTerm Subject  Number  \
6663  2022  Fall  2022-fa      IS     100   
6664  2022  Fall  2022-fa      IS     100   
6665  2022  Fall  2022-fa      IS     100   
6666  2022  Fall  2022-fa      IS     100   
6667  2022  Fall  2022-fa      IS     100   
...    ...   ...      ...     ...     ...   
6907  2022  Fall  2022-fa      IS     596   
6908  2022  Fall  2022-fa      IS     597   
6909  2022  Fall  2022-fa      IS     597   
6910  2022  Fall  2022-fa      IS     597   
6911  2022  Fall  2022-fa      IS     599   

                                                   Name  \
6663   Exploring the iSchool with a Human-Centered Lens   
6664   Exploring the iSchool with a Human-Centered Lens   
6665   Exploring the iSchool with a Human-Centered Lens   
6666   Exploring the iSchool with a Human-Centered Lens   
6667   Exploring the iSchool with a Human-Centered Lens   
...                                                 ...   
6907  Advanced Topics in Human-Centered Design &amp;...   
6908  Advanced Topics in Data Analytics &amp; Data S...   
6909  Advanced Topics in Data Analytics &amp; Data S...   
6910  Advanced Topics in Data Analytics &amp; Data S...   
6911                                    Thesis Research   

                                            Description    Credit Hours  \
6663  This course introduces students to the School ...        1 hours.   
6664  This course introduces students to the School ...        1 hours.   
6665  This course introduces students to the School ...        1 hours.   
6666  This course introduces students to the School ...        1 hours.   
6667  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
6907  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6908  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6909  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6910  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6911  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
6663  Prerequisite: Restricted to Majors Only; First...               NaN   
6664  Prerequisite: Restricted to Majors Only; First...               NaN   
6665  Prerequisite: Restricted to Majors Only; First...               NaN   
6666  Prerequisite: Restricted to Majors Only; First...               NaN   
6667  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
6907  2 or 4 graduate hours. No professional credit....               NaN   
6908  2 or 4 graduate hours. No professional credit....               NaN   
6909  2 or 4 graduate hours. No professional credit....               NaN   
6910  2 or 4 graduate hours. No professional credit....               NaN   
6911  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status           Enrollment Status                Type  \
6663  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6664  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6665  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6666  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6667  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
...   ...            ...                         ...                 ...   
6907  ...              A           Open (Restricted)              Online   
6908  ...              A           Open (Restricted)  Lecture-Discussion   
6909  ...              A           Open (Restricted)  Lecture-Discussion   
6910  ...              A           Open (Restricted)              Online   
6911  ...              A           Open (Restricted)   Independent Study   

     Type Code St

In [ ]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

*The IS program over the course of two years grew by 170.7%. During its initial year it had 92 courses which has now become 249 courses.*

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)